neo/Consensus/ConsensusService.cs

```diff
@@ -270,11 +274,30 @@ private void OnConsensusPayload(ConsensusPayload payload)
         private void OnPersistCompleted(Block block)
         {
             Log($"persist block: {block.Hash}");
-            block_received_time = TimeProvider.Current.UtcNow;
+            reference_block_time = TimeProvider.Current.UtcNow;        
+            theoreticalDelay = calcNetworkLatencyOrClockDriftWithMaxOffset(reference_block_time.ToTimestamp(), block.Timestamp);
+            reference_block_time.AddSeconds((double)-theoreticalDelay);
             knownHashes.Clear();
             InitializeConsensus(0);
         }
 
+        // Simple formula based on a filter considering timestamp of last block
+        // variable maxDelayToAdvance limits the offset adjustment to 30% of block time - currently 4.5s for 15s blocks
+        private uint calcNetworkLatencyOrClockDriftWithMaxOffset(uint localClockTimestamp, uint lastBlockTimestamp)
+        {
+            if (localClockTimestamp <= lastBlockTimestamp)
+                return 0;
+            
+            var currentTheoreticalDelay = localClockTimestamp - lastBlockTimestamp;
+            Log($"localClock:{localClockTimestamp}\nlastBlockTimestamp:{lastBlockTimestamp}");
+            Log($"diff:{currentTheoreticalDelay}");
+            double maxTimeoutOffset = Math.Ceiling(Blockchain.SecondsPerBlock * 0.3);
```

Comment:
Perhaps this method should just return a double, then you don't have to round this up, and you don't have to cast it to a double outside the method where it is used.

Reply:
Sounds good.